zfit/constraint.py

```diff
@@ -1,19 +1,53 @@
 from zfit import ztf
 from .util.container import convert_to_container
+import numpy as np
+import tensorflow as tf
 
 __all__ = ["nll_gaussian"]
 
+
 def nll_gaussian(params, mu, sigma):
-    params = convert_to_container(params, container=tuple)
-    mu = convert_to_container(mu, container=tuple)
-    sigma = convert_to_container(sigma, container=tuple)
-    constraint = ztf.constant(0.)
-    if not len(params) == len(mu) == len(sigma):
+    """Return negative log likelihood graph for gaussian constraints on a list of parameters.
+
+    Args:
+        params (list(zfit.Parameter)): The parameters to constraint
+        mu (float, list(float) or numpy array): The central value of the constraint
+        sigma (float, list(float) or numpy array): The standard deviations or covariance matrix of the 
+        constraint
+    Returns:
+        graph: the nll of the constraint
+    Raises:
+        ValueError: if params, mu and sigma don't have the same size
+    """
+
+    params = convert_to_container(params, tuple)
+    mu = convert_to_container(mu, tuple)
+
```

Comment:
The function should in general accept tensors, python and numpy objects being the a special case of a constant tensor.
This makes things of course a little bit more complicated :)
params and mu have to be scalars (probably allow mu as tensor later on), we leave it like that. But make sure that sigma can be a tensor instead of a numpy array. So let's do all the conversion to Tensor as soon as possible and use all tf operations (tf.diag) to do the rest.

This should also remove to special case for `isnumber`, since it's just the special case of a Tensor with n axes.

You agree? Let me know if you get stuck with TF somewhere